test/formal/signextend_shift.py

```diff
@@ -0,0 +1,43 @@
+from rule import Rule
+from opcodes import *
+
+"""
+Rule:
+1) SHL(A, SIGNEXTEND(B, X)) -> SIGNEXTEND(A / 8 + B, SHL(A, X))
+given A % 8 == 0 AND A <= 256 && B <= 32
+
+2) SIGNEXTEND(A, SHR(B, X)) -> SAR(B, X)
+given
+    B % 8 == 0 AND
+    A <= WordSize
+    B <= wordSize
+    (WordSize - B) / 8 == A + 1
+"""
+
+n_bits = 128
+
+# Input vars
+X = BitVec('X', n_bits)
+Y = BitVec('Y', n_bits)
+A = BitVec('A', n_bits)
+B = BitVec('B', n_bits)
+
+rule1 = Rule()
+rule1.require(A % 8 == 0)
+rule1.require(ULE(A, 256))
+rule1.require(ULE(B, 32))
+rule1.check(
+    SHL(A, SIGNEXTEND(B, X)),
+    SIGNEXTEND(A / 8 + B, SHL(A, X))
+)
```

Comment:
```suggestion
rule1 = Rule()
rule1.require(A & 7 == 0)
rule1.require(ULE(A, n_bits))
rule1.require(ULE(B, n_bits/8))
rule1.check(
    SHL(A, SIGNEXTEND(B, X)),
    SIGNEXTEND(LShR(A, 3) + B, SHL(A, X))
)
```
This way it proves easily for ``n_bits = 256`` bits as well... might be nicer to actually formulate the rule like that and then have full proofs... all others prove for be for 256 bits as well, but it can take a while.

Reply:
I actually thought that proving it for a different word size would be nice to show that we use "WordSize" correctly in the code.

About `&7`: If the code uses `%8`, why not use `%8` - same with `/8`.

Reply:
Why prove a different rule than the one we apply (as in a version with a different word size)? And as I said I'd suggest to change the actual rule to use masking and shifting instead of modulo and division. It's more efficient anyways and if that version can by verified why use a different version?